README.md
# [![Movable Titlebar Buttons, Firefox Addon][Logo] Movable Titlebar Buttons][Movable Titlebar Buttons, Firefox Addon]

[![License]][url: License] [![Semver]][url: Semver] [![Gittip]][url: Gittip] [![Download]][url: Download]

> Movable Titlebar Buttons, Firefox Addon, allows you to move the titlebar buttons on Windows 10 (only).

## Screenshots

Here are some screenshots from the addon:

### Together with [Line, Firefox Addon][Line, Firefox Addon]

![Together with Line, Firefox Addon](https://cdn.pbrd.co/images/efItPt1.png)

This is highly recommended as it creates a super minimalistic experience.

### Standalone

![Basic functionality](https://cdn.pbrd.co/images/efRxEqJ.png)

# Features

+ Allows you to move the minimize, maximize/restore, close buttons in Firefox on Windows 10. These buttons can be placed anywhere you can customize it to.

## Installation & Building

### Requirements

+ `Firefox 38+`

### [From AMO (Mozilla Addons)](https://addons.mozilla.org/en-US/firefox/addon/movable-titlebar-buttons/)

Note: Not yet signed, pre-release!

### From Github [![Download]][url: Download]

### Manually, from repository

First clone the repository to your local machine:

```shell
git clone https://github.com/Centril/firefox-movable-titlebar && cd firefox-movable-titlebar
```

Since this is written with the **[Addon SDK]**, you will need to use **[`jpm`]**.

If you haven't installed it before, simply do (you might need to run this with `sudo`):

```shell
npm install jpm -g
```

To test out the addon, try:

```shell
jpm run
```

If you want to package it as an `.xpi`, do:

```shell
jpm xpi
```

## Changelog

See **[CHANGES.md]**.

## Copyright & License

Licensed under the **[GPL 2 License]**.
Copyright 2015 Mazdak Farrokhzad for the modified parts.

## Bugs | Issues | Feature requests | Contribution

Want to contribute? Great stuff! Please use the issue system that github provides to report bugs/issues or request an enhancement. Pull requests are also more than welcome.

## Author

**Mazdak Farrokhzad / Centril [&lt;twingoow@gmail.com&gt;]**

[![twitter][twitter_image]][twitter] [![github][github_image]][github] [![facebook][facebook_image]][facebook]

<!-- references -->

[Gittip]: http://img.shields.io/gittip/Centril.svg?style=flat
[url: Gittip]: https://www.gittip.com/Centril/
[License]: http://img.shields.io/badge/license-GPL_2+-blue.svg?style=flat
[url: License]: LICENSE.md
[Semver]: http://img.shields.io/badge/semver-2.0.0-blue.svg?style=flat
[url: Semver]: http://semver.org/spec/v2.0.0.html
[Download]: https://img.shields.io/badge/Download_XPI-1.0.1--pre-ff69b4.svg?style=flat
[url: Download]: https://github.com/Centril/firefox-movable-titlebar/releases/tag/1.0.1-pre

[Logo]: https://raw.githubusercontent.com/Centril/firefox-movable-titlebar/master/icon.png
[Movable Titlebar Buttons, Firefox Addon]: https://github.com/Centril/firefox-movable-titlebar
[Line, Firefox Addon]: https://github.com/Centril/firefox-line

[Addon SDK]: https://developer.mozilla.org/en-US/Add-ons/SDK
[`jpm`]: https://developer.mozilla.org/en-US/Add-ons/SDK/Tools/jpm#Installation

[twitter]: http://twitter.com/CenoRIX
[twitter_image]: http://cdn.flaticon.com/png/128/8800.png
[github]: https://github.com/centril
[github_image]: http://cdn.flaticon.com/png/128/25231.png
[facebook]: https://www.facebook.com/Centril
[facebook_image]: http://cdn.flaticon.com/png/128/33702.png
[&lt;twingoow@gmail.com&gt;]: mailto:twingoow@gmail.com

[CHANGES.md]: CHANGES.md
[GPL 2 License]: LICENSE.md

<!-- references -->

CHANGES.md
# CHANELOG

## 1.0.1

+ Initial release. Allows you to move the minimize, maximize/restore, close buttons in Firefox on Windows 10. These buttons can be placed anywhere you can customize it to.

src/ids.js
/*
 * Movable Titlebar Buttons, a Firefox Addon/Extension,
 * allows you to move the titlebar buttons on Windows 10.
 *
 * Copyright (C) 2015 Mazdak Farrokhzad <twingoow@gmail.com>, Mozilla Labs
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
'use strict';

// Define all IDs used in addon:
exports.ID = {
	replacement: {
		min:	'firefox-movable-titlebar-min',
		max:	'firefox-movable-titlebar-max',
		close:	'firefox-movable-titlebar-close'
	}
};

src/utils.js
/*
 * Movable Titlebar Buttons, a Firefox Addon/Extension,
 * allows you to move the titlebar buttons on Windows 10.
 *
 * Copyright (C) 2015 Mazdak Farrokhzad <twingoow@gmail.com>, Mozilla Labs
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
'use strict';

// -------------------------------------------------------------------------
// SDK & JSM Loading:
// -------------------------------------------------------------------------

/**
 * Returns a list of SDKs.
 *
 * @param  {String[]}  list  A string list of sdks without 'sdk/' path prefix.
 * @return {Object[]}        The SDKs.
 */
const sdks = list => list.map( v => require( `sdk/${v}` ) );
exports.sdks = sdks;

const requireJSM = jsm => require( `resource://${jsm}.jsm` );
exports.requireJSM = requireJSM;

// -------------------------------------------------------------------------
// Load everything this unit needs:
// -------------------------------------------------------------------------

// Get CUI, Services:
const { CustomizableUI: CUI } = requireJSM( '/modules/CustomizableUI' );
exports.CUI = CUI;

const [ window_utils, {browserWindows: windows},
		{viewFor}, {when: unloader}, {partial, delay},
		{Class: _class, mix}, {attachTo, detachFrom}, {Style},
		{isArray}] = sdks(
	  [ 'window/utils', 'windows',
	    'view/core', 'system/unload', 'lang/functional',
	    'core/heritage', 'content/mod', 'stylesheet/style',
	    'lang/type'] );

// -------------------------------------------------------------------------
// General purpose functions:
// -------------------------------------------------------------------------

const each = (obj, fn) => {
	if ( isArray( obj ) ) return obj.forEach( fn );
	Object.keys( obj ).forEach( k => fn( k, obj[k] ) );
};
exports.each = each;

// -------------------------------------------------------------------------
// Windows:
// -------------------------------------------------------------------------

/**
 * Executes fn, passed args and
 * registers unload that is called with args on unload.
 *
 * @param  {Function}  fn     The function to call.
 * @param  {Function}  unload The function to unload what fn did with.
 * @param  {...*} args        The arguments to pass to fn & unload.
 */
const unloadable = (fn, unload, ...args) => {
	fn( ...args );
	unloader( partial( unload, args ) );
};
exports.unloadable = unloadable;

/**
 * Apply a callback to each open and new browser windows.
 *
 * @usage watchWindows( callback ): Apply a callback to each browser window.
 * @param {function} callback: 1-parameter function that gets a browser window.
 */
const watchWindows = callback => {
	// Add functionality to existing windows
	each( window_utils.windows( 'navigator:browser', {includePrivate: true} ), callback );

	// Watch for new browser windows opening:
	unloadable( l => windows.on( 'open', l ), l => windows.off( 'open', l ),
		w => callback( viewFor( w ) ) );
}
exports.watchWindows = watchWindows;

/**
 * Asynchronously apply a callback to each open and new browser windows.
 *
 * @usage watchWindowsAsync( callback ): Apply a callback to each browser window.
 * @param {function} callback: 1-parameter function that gets a browser window.
 */
const watchWindowsAsync = callback =>
	watchWindows( window => delay( () => callback( window ), 0 ) );
exports.watchWindowsAsync = watchWindowsAsync;

// -------------------------------------------------------------------------
// CUI:
// -------------------------------------------------------------------------

/**
 * Executes fn in a CUI batch update.
 *
 * @param  {Function} fn The function to execute.
 * @return {*}           Whatever fn returns.
 */
const cuiDo = (fn, ...args) => {
	try {
		CUI.beginBatchUpdate();
		return fn( ...args );
	} finally {
		CUI.endBatchUpdate();
	}
};
exports.cuiDo = cuiDo;

// -------------------------------------------------------------------------
// WindowFeature:
// -------------------------------------------------------------------------

/**
 * WindowFeature, a helper for a per window addon feature:
 */
const WindowFeature = _class( {
	/**
	 * Constructor:
	 *
	 * @param  {ChromeWindow} window The Window.
	 */
	initialize( window ) {
		this.window = window;
		this.doc = window.document;
	},

	/**
	 * Applies the style given by uri.
	 *
	 * @param  {String} uri the URI of the style CSS file.
	 * @return {Style}      the style attachment, to unload later.
	 */
	style( uri ) {
		return this.attach( new Style( { uri: uri } ) );
	},

	/**
	 * Attaches a modification to our window.
	 *
	 * @param  {*} mod  The modification to attach.
	 * @return {*}      The argument mod.
	 */
	attach( mod ) { attachTo( mod, this.window ); return mod; },

	/**
	 * Detaches a modification from our window.
	 *
	 * @param  {*} mod  The modification to detach.
	 * @return {null}   Null.
	 */
	detach( mod ) { detachFrom( mod, this.window ); return null; },
} );

WindowFeature.extend = proto => _class( mix( proto, { extends: WindowFeature } ) );

exports.WindowFeature = WindowFeature;

/**
 * Watch a WindowFeature, applying it async as windows come.
 *
 * @param  {WindowFeature} f The WindowFeature, expecting a function apply on prototype.
 */
const watchFeature = f => watchWindowsAsync( window => f( window ).apply() );
exports.watchFeature = watchFeature;

src/main.js
/*
 * Movable Titlebar Buttons, a Firefox Addon/Extension,
 * allows you to move the titlebar buttons on Windows 10.
 *
 * Copyright (C) 2015 Mazdak Farrokhzad <twingoow@gmail.com>, Mozilla Labs
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
'use strict';

// Import utils, ids, SDK:
const { ID }							= require( './ids' );
const {	sdks, each, CUI, cuiDo,
		watchFeature, WindowFeature }	= require('./utils');
const [ {when: unloader} ]				= sdks( ['system/unload'] );

const addButton = action => CUI.createWidget( {
	type: 'button',
	showInPrivateBrowsing: true,
	id: action[0],
	onCommand: evt => action[1]( evt.target.ownerDocument.defaultView ),
	label: ' ',
	tooltiptext: ' '
	// @TODO add labels if anyone asks for them...
} );

const destroyButton = widget => CUI.destroyWidget( widget.id );

// Apply to each window as they come:
watchFeature( WindowFeature.extend( { apply() {
	// Apply movable-titlebar.css:
	const style = this.style( './movable-titlebar.css' );

	// Hide the originals:
	const root = this.doc.documentElement;
	root.setAttribute( 'movable-titlebar', true );

	// Clean up various changes when the add-on unloads:
	unloader( () => {
		// Restore originals:
		root.removeAttribute( 'movable-titlebar' );

		// Remove our style:
		this.detach( style );
	} );
} } ) );

// Create new buttons & unload them when needed:
const R = ID.replacement;
const news = cuiDo( news => news.map( addButton ),
	[[R.min,	window => window.minimize()],
	 [R.max,	window => window.onTitlebarMaxClick()],
	 [R.close,	window => window.BrowserTryToCloseWindow()]] );
unloader( () => each( news, destroyButton ) );